pkg/controller/application/application_controller.go

```diff
@@ -312,11 +312,17 @@ func (c *Controller) wrapUpApplicationConditions(app *shipper.Application, rels
 	rels = releaseutil.SortByGenerationDescending(rels)
 
 	abortingCond := apputil.NewApplicationCondition(shipper.ApplicationConditionTypeAborting, corev1.ConditionFalse, "", "")
-	apputil.SetApplicationCondition(&app.Status, *abortingCond)
+	if diff := apputil.SetApplicationCondition(&app.Status, *abortingCond); !diff.IsEmpty() {
+		c.reportApplicationConditionChange(app, diff)
+	}
 	validHistoryCond := apputil.NewApplicationCondition(shipper.ApplicationConditionTypeValidHistory, corev1.ConditionTrue, "", "")
-	apputil.SetApplicationCondition(&app.Status, *validHistoryCond)
+	if diff := apputil.SetApplicationCondition(&app.Status, *validHistoryCond); !diff.IsEmpty() {
+		c.reportApplicationConditionChange(app, diff)
```

Comment:
This seems a bit repetitive. What do you think about having a set of diffs in the same struct instead? Then we could check it only once at the end, and apply the relevant operations.